src/utilities/map.ts

```diff
@@ -0,0 +1,33 @@
+/**
+ *
+ * Adds an element to the array associated with a key, or creates a new array if one does not exist yet.
+ *
+ * @param key The key of the value the value which will be added
+ * @param value The value that will be added to the array
+ */
+export function mapAdd<T>(map: Map<string, T[]>, key: string, value: T) {
+    if (map.has(key)) {
+        const entry = map.get(key);
+        if (entry !== undefined) {
+            entry.push(value);
+            map.set(key, entry);
+        }
+    } else {
+        map.set(key, [value]);
+    }
+}
```

Comment:
Why is this an impure function? Why do we need it? map has methods for this built in iirc?